### PR TITLE
introduce SinkType[T], SinkFn[T] and ErrInvalidSink 

### PR DESCRIPTION
**Pull request by amery**
## Summary by CodeRabbit

- **New Features**
	- Introduced `ErrInvalidSink` for improved error messaging related to sink usability.
	- Added a new `SinkType` struct for creating and managing sink objects with enhanced validation and error handling.
	- Implemented methods for encoding, decoding, and managing data within sink objects, ensuring data integrity and expiration management.

### CODE REVIEW

File: sink_fn.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	if !sink.Valid() {
+		return nil, false
+	}
+
+	// copy
+	if sink.val != nil {
+		out, ok := DefaultClone(sink.val)
+		if ok {
+			return out, true
+		}
```
[the remark this answers is not in the document]
Comment (karasz) [reply sink_fn.go:142]: I agree on this one with @coderabbitai

[the remark this answers is not in the document]
Comment (amery) [reply sink_fn.go]: amended

File: sink_fn.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	_ TSink[any] = (*SinkFn[any])(nil)
+)
+
+// SinkFn is a [TSink] implementation made using a [SinkType] factory.
+type SinkFn[T any] struct {
+	ByteSink
+
+	typ *SinkType[T]
+	val *T
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply sink_fn.go:53]: Linters hate public methods returning private types

**Review by karasz — APPROVED** [review]
ok

**Comment by amery** [comment]
`darvaza.org/cache@v0.3.1` tagged

**Pull request merged**